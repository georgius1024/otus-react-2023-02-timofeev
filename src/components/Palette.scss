.palette {
  display: flex;
  justify-content: center;
  .sample {
    cursor: pointer;
    margin: 0 12px;
    outline: 1px solid currentColor;
    outline-offset: 1px;
    border-radius: 100%;
    display: flex;
    user-select: none;
    align-items: center;
    justify-content: center;
    transition: 200ms ease;
    outline-offset: 2px;
    outline-width: 4px;
    outline-color: #333;
    outline-style: none;

    &:hover {
      background-color: #aaa;
      transform: scale(1.2);
      filter: drop-shadow(4px 4px 4px #777);
    }
  }
}
