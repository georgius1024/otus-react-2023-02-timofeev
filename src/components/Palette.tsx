import "@/components/Palette.scss";
import { useState, ReactElement, useCallback } from "react";
type Sample = {
  color: string;
  symbol: string;
};
type clickHandler = (sample: Sample | null) => void;

type PaletteProps = {
  onClick: clickHandler;
};

const NODE_SIZE = 48;
const NODE_SYMBOLS = ["🐷", "🐱", "🐶", "🐭", "🦊", "🐻", "❌"];
const NODE_COLORS = [
  "red",
  "silver",
  "navy",
  "orange",
  "green",
  "teal",
  "white",
];

const shuffle = (list: string[]) => list.sort(Math.random);

export default function Palette(props: PaletteProps): ReactElement {
  const [selected, setSelected] = useState<number>(0);

  const sampleStyle = (sample: Sample, index: number) => ({
    width: `${NODE_SIZE}px`,
    height: `${NODE_SIZE}px`,
    fontSize: `${NODE_SIZE * 0.66}px`,
    color: sample.color,
    backgroundColor: sample.color,
    outlineStyle: index === selected ? "solid" : "none",
  });

  const sampleClickHandler = useCallback(
    (sample: Sample, index: number) =>
      (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
        event.stopPropagation();
        event.preventDefault();
        if (sample.symbol !== "❌") {
          props.onClick(sample);
          setSelected(index);
        } else {
          props.onClick(null);
          setSelected(-1);
        }
      },
    []
  );

  const samples = NODE_SYMBOLS.map((symbol, index): Sample => {
    return {
      symbol,
      color: NODE_COLORS[index],
    };
  }).map((sample: Sample, index: number) => {
    return (
      <div
        className="sample"
        key={sample.symbol}
        style={sampleStyle(sample, index)}
        onClick={sampleClickHandler(sample, index)}
      >
        {sample.symbol}
      </div>
    );
  });
  return <div className="palette">{samples}</div>;
}
