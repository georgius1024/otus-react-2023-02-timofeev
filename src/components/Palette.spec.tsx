import { render, screen } from "@testing-library/react";
import { vi, expect } from "vitest";
import Palette from "@/components/Palette";

describe("Palette", () => {
  const clickHandler = vi.fn();
  let wrapper: any
  beforeEach(() => {
    wrapper = render(
      <Palette
        onClick={clickHandler}
      />
    );
  });
  afterEach(() => {
    clickHandler.mockReset();
  });
  it("renders without errors", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("dot clicks emits events", async () => {
    const Palette = await screen.findByText('‚ùå');
    expect(Palette).toBeVisible();
    Palette.click();
    expect(clickHandler).toHaveBeenCalled();
  });
});
