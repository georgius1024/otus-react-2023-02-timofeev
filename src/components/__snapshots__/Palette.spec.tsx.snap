// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Palette > renders without errors 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="palette"
      >
        <div
          class="sample"
          style="width: 48px; height: 48px; font-size: 31.68px; color: red; background-color: red; outline-style: solid;"
        >
          üê∑
        </div>
        <div
          class="sample"
          style="width: 48px; height: 48px; font-size: 31.68px; color: silver; background-color: silver; outline-style: none;"
        >
          üê±
        </div>
        <div
          class="sample"
          style="width: 48px; height: 48px; font-size: 31.68px; color: navy; background-color: navy; outline-style: none;"
        >
          üê∂
        </div>
        <div
          class="sample"
          style="width: 48px; height: 48px; font-size: 31.68px; color: orange; background-color: orange; outline-style: none;"
        >
          üê≠
        </div>
        <div
          class="sample"
          style="width: 48px; height: 48px; font-size: 31.68px; color: green; background-color: green; outline-style: none;"
        >
          ü¶ä
        </div>
        <div
          class="sample"
          style="width: 48px; height: 48px; font-size: 31.68px; color: teal; background-color: teal; outline-style: none;"
        >
          üêª
        </div>
        <div
          class="sample"
          style="width: 48px; height: 48px; font-size: 31.68px; color: white; background-color: white; outline-style: none;"
        >
          ‚ùå
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="palette"
    >
      <div
        class="sample"
        style="width: 48px; height: 48px; font-size: 31.68px; color: red; background-color: red; outline-style: solid;"
      >
        üê∑
      </div>
      <div
        class="sample"
        style="width: 48px; height: 48px; font-size: 31.68px; color: silver; background-color: silver; outline-style: none;"
      >
        üê±
      </div>
      <div
        class="sample"
        style="width: 48px; height: 48px; font-size: 31.68px; color: navy; background-color: navy; outline-style: none;"
      >
        üê∂
      </div>
      <div
        class="sample"
        style="width: 48px; height: 48px; font-size: 31.68px; color: orange; background-color: orange; outline-style: none;"
      >
        üê≠
      </div>
      <div
        class="sample"
        style="width: 48px; height: 48px; font-size: 31.68px; color: green; background-color: green; outline-style: none;"
      >
        ü¶ä
      </div>
      <div
        class="sample"
        style="width: 48px; height: 48px; font-size: 31.68px; color: teal; background-color: teal; outline-style: none;"
      >
        üêª
      </div>
      <div
        class="sample"
        style="width: 48px; height: 48px; font-size: 31.68px; color: white; background-color: white; outline-style: none;"
      >
        ‚ùå
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
